crates/wasmtime/src/runtime/vm/continuation.rs

```diff
@@ -1,459 +1,320 @@
-/// `StackLimits` object contains stale/undefined data. Instead, the live data
-/// describing the limits for the currently executing stack is always maintained
-/// in `VMRuntimeLimits`. Note that as a general rule independently from any
-/// execution of continuations, the `last_wasm_exit*` fields in the
-/// `VMRuntimeLimits` contain undefined values while executing wasm.
-///
-/// Parents of currently executing stack:
-/// For stacks that appear in the tail of the VMContext's
-/// `typed_continuations_chain` list (i.e., stacks that are not currently
-/// executing themselves, but are a parent of the currently executing stack), we
-/// have the following: All the fields in the stack's StackLimits are valid,
-/// describing the stack's stack limit, and pointers where executing for that
-/// stack entered and exited WASM.
-///
-/// Suspended continuations:
-/// For suspended continuations (including their parents), we have the
-/// following. Note that the main stack can never be in this state. The
-/// `stack_limit` and `last_enter_wasm_sp` fields of the corresponding
-/// `StackLimits` object contain valid data, while the `last_exit_wasm_*` fields
-/// contain arbitrary values.
-/// There is only one exception to this: Note that a continuation that has been
-/// created with cont.new, but never been resumed so far, is considered
-/// "suspended". However, its `last_enter_wasm_sp` field contains undefined
-/// data. This is justified, because when resume-ing a continuation for the
-/// first time, a native-to-wasm trampoline is called, which sets up the
-/// `last_wasm_entry_sp` in the `VMRuntimeLimits` with the correct value, thus
-/// restoring the necessary invariant.
-#[derive(Debug, Clone, PartialEq)]
-#[repr(usize, C)]
-pub enum StackChain {
-    /// If stored in the VMContext, used to indicate that the MainStack entry
-    /// has not been set, yet. If stored in a VMContRef's parent_chain
-    /// field, means that there is currently no parent.
-    Absent = wasmtime_continuations::STACK_CHAIN_ABSENT_DISCRIMINANT,
-    /// Represents the main stack.
-    MainStack(*mut StackLimits) = wasmtime_continuations::STACK_CHAIN_MAIN_STACK_DISCRIMINANT,
-    /// Represents a continuation's stack.
-    Continuation(*mut VMContRef) = wasmtime_continuations::STACK_CHAIN_CONTINUATION_DISCRIMINANT,
+cfg_if::cfg_if! {
+    if #[cfg(feature = "wasmfx_baseline")] {
+        pub use baseline as imp;
+    } else {
+        pub use optimized as imp;
+    }
 }
 
-impl StackChain {
-    /// Indicates if `self` is a `MainStack` variant.
-    pub fn is_main_stack(&self) -> bool {
-        matches!(self, StackChain::MainStack(_))
-    }
+/// M:1 Many-to-one mapping. A single VMContRef may be
+/// referenced by multiple VMContObj, though, only one
+/// VMContObj may hold a non-null reference to the object
+/// at a given time.
+#[repr(C)]
+pub struct VMContObj(pub Option<*mut imp::VMContRef>);
+
+#[cfg(not(feature = "wasmfx_baseline"))]
```

Comment:
Can we please _compile_ both `baseline` and `optimized` unconditionally? When working on refactorings it's very inconvenient if you can only type-check one of the implementations at a time and need to enable/disable the corresponding feature, which involves recompiling a lot of `wasmtime`.
We could then ensure that only the "active" implementation is actually `pub` and/or redirect the inactive one to the dummy modules you define further down.

Reply:
Yes, but not in this PR. We can do that at in later PR, as it would require defining a common interface.